### PR TITLE
[FIXED JENKINS-32571] Preserve original security settings after calling CommandInvoker.invoke()

### PR DESCRIPTION
**Pull request by pjanouse**
[JENKINS-32571](https://issues.jenkins-ci.org/browse/JENKINS-32571)

Please review. Thx Pavel

### CODE REVIEW

**Comment by pjanouse** [comment]
PR#4 should fix the build issue.

File: src/main/java/hudson/cli/CLICommandInvoker.java [hunk 1]
```diff
@@ -121,7 +128,11 @@ public Result invoke() {
                 args, locale, stdin, new PrintStream(out), new PrintStream(err)
         );
 
-        return new Result(returnCode, out, err);
+        Result result = new Result(returnCode, out, err);
+
+        restoreAuth();
```
Comment (jglick) [inline src/main/java/hudson/cli/CLICommandInvoker.java]: A little safer to do this in a `finally` block. `main` does catch `Exception`, though, so probably good enough (any `Error` is most likely going to fail the whole test case anyway).

**Comment by jglick** [comment]
:+1:

**Comment by jglick** [comment]
`git merge master && git push` to have CI check this.

**Comment by pjanouse** [comment]
I've fixed calling the restoreAuth().

I'd like to explain more what is my intention in the a bit longer code. First - I'd like to distinct an error occurred in command.main(...) from the error occurred during restoreAuth().

If an issue occurred in the finally block (while restoreAuth() performed) and I'll only catch and report that, the original issue from command.main(...) (and probably an inconsistent state of system) is hidden.

The second option is don't catch an exception while performing restoreAuth() at all and left everything on the caller. On the other hand in that case the root cause can be overlooked and interchanged with the dependent one.

WDYT about this approach?

**Comment by jglick** [comment]
> the original issue from `command.main(...)` is hidden

So? If there is an error in `restoreAuth`, the test is busted and needs to be fixed anyway.

(Java 7+ has a suppressed exception facility, though I think it is overkill here.)

File: src/main/java/hudson/cli/CLICommandInvoker.java [hunk 2]
```diff
@@ -112,16 +120,35 @@ public Result invokeWithArgs(final String... args) {
...
+            executionError = e;
+            throw e;
+        } finally {
+            try {
+                restoreAuth();
+            } catch (Exception e) {
+                if (executionError != null) { // Log original Error is occurred
+                    throw new AssertionError("Restore authentization failed and execution failure occurred previously.", executionError);
+                } else {
+                    throw new AssertionError("Restore authentization failed.");
```
Comment (jglick) [inline src/main/java/hudson/cli/CLICommandInvoker.java]: Now (in either case) you are losing `e`.

**Comment by jglick** [comment]
Seems OK as is, though personally I would just simplify to

```java
try {
    setAuth();
    return …;
} finally {
    restoreAuth();
}
```

**Comment by pjanouse** [comment]
Jesse,

I'm on the same page with you on "do a thing as simple as we can" however I can see two different points of view in this case:
- developer PoV
- test executor PoV

Your approach can be the best for the second one - if something went wrong, it doesn't matter so much what/where an issue occurred, sth. is broken and has to be fixed; period. On the other hand a developer can appreciate if he knows the root cause precisely, in the first shot and without any confusion. 

I see the CLICommandInvoker class primary as a helper for a developer - easily and rapidly develop a CLI test so as a developer I don't like the result produced by your version in this scenario:
- called invoke()
- exception is thrown in command.main()
- exception is thrown in restoreAuth()

In above case the root exception is eaten and JUNIT/Surefire reports only the second one. If the first exception is visible/reported it depends on the impl. code/logging, in many common cases it needn't to be reported anywhere and developer can be completely confused why such an exception was thrown.

Do you really think the above example is so rare exception that we don't need to deal it at all? I'm still inclining to solve it. If you have an idea how to implement that in a better way, please share your thought.

**Comment by pjanouse** [comment]
I've thought about this a bit more, what about the original approach - not to envelope the code with try/catch block at all... if exception is thrown in command.main() fine, will be reported and further execution is unclean; it doesn't matter so much if we call restoreAuth() anyway. If the exception is thrown in restoreAuth(), it will be reported as well.

I mean code like: 
```java
public invoke() {
  setAuth();
  command.main();
  restoreAuth();
}
```

I think that this simple approach solves my objection. WDYT?

**Comment by pjanouse** [comment]
I've returned to my original impl. which fixes above issue which I'd like to solve as well.

**Comment by pjanouse** [comment]
Could someone review and if no objection merge this change please?

It effectively block e.g. JENKINS-32535.

Thank you, Pavel

File: src/main/java/hudson/cli/CLICommandInvoker.java [hunk 3]
```diff
@@ -112,6 +120,8 @@ public Result invokeWithArgs(final String... args) {
 
     public Result invoke() {
 
+        Result result;
+        Error executionError = null;
```
Comment (jglick) [inline src/main/java/hudson/cli/CLICommandInvoker.java]: Seems that `executionError` is unused, no?

Also I see no reason why `result` is being defined here. Would seem simpler for end of method to read

```java
restoreAuth();
return new Result(returnCode, out, err);
```

**Comment by jglick** [comment]
> Do you really think the above example is so rare exception that we don't need to deal it at all?

Yes probably it is very rare.

BTW in `-source 7` you could make `setAuth` return some `AutoCloseable` and then use `try`-with-resources, or simply call `Throwable.addSuppressedException` directly. Anyway this is all probably overkill for such a minor bit of test code.

**Pull request merged**